pkg/auth/requests/service_account.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/steve/pkg/auth"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	"k8s.io/client-go/rest"
+)
+
+const serviceAccountSubjectPrefix = "system:serviceaccount:"
+
+type restConfigGetter func(cluster *mgmtv3.Cluster, context *config.ScaledContext, secretLister corev1.SecretLister) (*rest.Config, error)
+
+// ServiceAccountAuth is an authenticator that authenticates requests using the downstream service account's JWT.
+type ServiceAccountAuth struct {
+	scaledContext    *config.ScaledContext
+	clusterLister    mgmtv3.ClusterLister
+	secretLister     corev1.SecretLister
+	restConfigGetter restConfigGetter
+}
+
+// NewServiceAccountAuth creates a new instance of ServiceAccountAuth.
+func NewServiceAccountAuth(
+	scaledContext *config.ScaledContext,
+	restConfigGetter restConfigGetter,
+) auth.Authenticator {
+	return &ServiceAccountAuth{
+		scaledContext:    scaledContext,
+		clusterLister:    scaledContext.Management.Clusters("").Controller().Lister(),
+		secretLister:     scaledContext.Core.Secrets("").Controller().Lister(),
+		restConfigGetter: restConfigGetter,
+	}
+}
+
+// Authenticate the request using the downstream service account's JWT.
+func (t *ServiceAccountAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	info, hasAuth := request.UserFrom(req.Context())
+	if info.GetName() != "system:cattle:error" {
+		return info, hasAuth, nil
+	}
+
+	// See if the token is a JWT.
+	rawToken := tokens.GetTokenAuthFromRequest(req)
+
+	jwtParser := jwtv4.Parser{}
+	claims := jwtv4.RegisteredClaims{}
+	_, _, err := jwtParser.ParseUnverified(rawToken, &claims)
+	if err != nil {
+		logrus.Debugf("saauth: error parsing JWT: %v", err)
+		return info, false, err
+	}
+
+	if !strings.HasPrefix(claims.Subject, serviceAccountSubjectPrefix) {
+		logrus.Debugf("saauth: JWT sub is not a service account: %v", err)
+		return info, false, nil
+	}
+
+	if isTokenExpired(claims.ExpiresAt) {
+		logrus.Debugf("saauth: Service Account JWT is expired. Expiration time was: %v", claims.ExpiresAt)
+		return info, false, nil
+	}
+
+	// Make sure the cluster exists.
+	clusterID := mux.Vars(req)["clusterID"]
+	if clusterID == "" {
+		return info, hasAuth, nil
+	}
+
+	cluster, err := t.clusterLister.Get("", clusterID)
+	if err != nil {
+		logrus.Debugf("saauth: error getting cluster: %v", err)
+		return info, false, err
+	}
+
+	// Get rest config for the cluster and instantiate an authentication client.
+	kubeConfig, err := t.restConfigGetter(cluster, t.scaledContext, t.secretLister)
```

Comment:
We may want to cache cluster REST configs. If we do we need to come up with solid invalidation strategy. E.g. refetch the config and retry if the auth client instantiation or a tokenreview call fails.